library/src/blas2/rocblas_syr.hpp

```diff
@@ -5,35 +5,79 @@
 #include "handle.h"
 #include "rocblas.h"
 
-template <typename T, typename U>
-__global__ void rocblas_syr_kernel(rocblas_fill uplo,
-                                   rocblas_int  n,
-                                   U            alpha_device_host,
-                                   const T* __restrict__ x,
-                                   rocblas_int incx,
-                                   T*          A,
-                                   rocblas_int lda)
+template <typename T, typename U, typename V, typename W>
+__global__ void rocblas_syr_kernel(rocblas_fill   uplo,
+                                   rocblas_int    n,
+                                   U              alpha_device_host,
+                                   rocblas_stride stride_alpha,
+                                   V              xa,
+                                   ptrdiff_t      shiftx,
+                                   rocblas_int    incx,
+                                   rocblas_stride stridex,
+                                   W              Aa,
+                                   ptrdiff_t      shiftA,
+                                   rocblas_int    lda,
+                                   rocblas_stride strideA)
```

Comment:
You don't need the template parameter T. You are using auto for x but T for A, mind as well use auto for A ?

Reply:
Yes I thought about that but I prefer to see the base type as a template.  This helps tracking the base type templates throughout the code starting from the API down to the kernel.   This is of course a personal preference, I also avoiding auto is good.

Reply:
How does it help tracking the base type? 
This abstraction has been made most of the time, auto is used especially on this purpose when loading batch ptr.

Reply:
I think it is a better design pattern to keep the core template type explicit, you see the template functions stand out to the reader as you read through the function.   Using auto here to save refactoring cost I don't buy that argument in this case.     ( code << refactor << read ) by orders of magnitude so this personally I think removing a typename by using auto is less readable.   Of course this is up for debate, but like your proposal to make the typename indicate something (e.g. S for scalar, or what I did in one function XPtr ) it might make the code more readable.    An open source library should be very readable if we want people to extend it, so I am suggesting we lean toward readability over minor code size reduction.   Sure the compiler can figure things out but often a reader has to figure something out too.   We can continue this discussion on Monday over coffee.

Reply:
I'm late to this conversation, but I think that not requiring explicit template type arguments (having them deduced instead) is preferred. `decltype()` can be used to "compute" the type based on another expression, without requiring the user to explicitly specify the type.

I don't understand: "Yes I thought about that but I prefer to see the base type as a template". If you're talking about `rocblas_float_complex` -> `float` and `rocblas_double_complex` -> `double`, then I think traits are a better way to do this -- in one of the header files, define a `real_type` template:

```
template<typename>
struct rocblas_real_type
{
};

template<>
struct rocblas_real_type<rocblas_float_complex>
{
    using type = float;
};

template<>
struct rocblas_real_type<rocblas_double_complex>
{
    using type = double;
};

using T = typename rocblas_real_type<U>::type;
```

Reply:
Actually, it would be simpler to just say:
```
using T = decltype(std::real(X));
```

i.e., use `decltype()` with `std::real` or a similar expression.

Reply:
Will resolve with a later PR as preference based.  Resolving this conversation.

Reply:
> I'm late to this conversation, but I think that not requiring explicit template type arguments (having them deduced instead) is preferred. `decltype()` can be used to "compute" the type based on another expression, without requiring the user to explicitly specify the type.
> 
> I don't understand: "Yes I thought about that but I prefer to see the base type as a template". If you're talking about `rocblas_float_complex` -> `float` and `rocblas_double_complex` -> `double`, then I think traits are a better way to do this -- in one of the header files, define a `real_type` template:
> 
> ```
> template<typename>
> struct rocblas_real_type
> {
> };
> 
> template<>
> struct rocblas_real_type<rocblas_float_complex>
> {
>     using type = float;
> };
> 
> template<>
> struct rocblas_real_type<rocblas_double_complex>
> {
>     using type = double;
> };
> 
> using T = typename rocblas_real_type<U>::type;
> ```

I came with a document on these topics.